.github/workflows/rust.yml

```diff
@@ -2,39 +2,137 @@ name: Rust
+            profile: minimal
+            toolchain: nightly
+            override: true
+            components: rustfmt
+
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@v1.3.0
 
+      - name: Cargo fmt
+        uses: actions-rs/cargo@v1.0.3
+        with:
+          command: fmt
+          args: --all -- --check
+
+  docs:
+    name: Check documentation
     runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v2
+
+      - name: Install Rust stable toolchain
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@v1.3.0
+
+      - name: Check internal documentation links
+        run: RUSTDOCFLAGS="--deny broken_intra_doc_links" cargo doc --verbose --workspace --no-deps --document-private-items
 
+  tests:
+    name: Cargo test
+    runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
+      - name: Checkout sources
+        uses: actions/checkout@v2
+
+      - name: Download Substrate
+        run: |
+            curl "https://releases.parity.io/substrate/x86_64-debian:stretch/latest/substrate/substrate" --output substrate --location
+            chmod +x ./substrate
+            mkdir -p ~/.local/bin
+            mv substrate ~/.local/bin
 
-    - name: setup
-      uses: actions-rs/toolchain@v1
-      with:
+      - name: Install Rust stable toolchain
+        uses: actions-rs/toolchain@v1
+        with:
           profile: minimal
-          toolchain: nightly
+          toolchain: stable
           override: true
-          components: rustfmt
-          target: wasm32-unknown-unknown
 
-    - name: download-substrate
-      run: |
-          curl "https://releases.parity.io/substrate/x86_64-debian:stretch/latest/substrate/substrate" --output substrate --location
-          chmod +x ./substrate
-          mkdir -p ~/.local/bin
-          mv substrate ~/.local/bin
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@v1.3.0
+
+      - name: Cargo test
+        uses: actions-rs/cargo@v1.0.3
+        with:
+          command: test
+          args: --all-targets --workspace
+
+  clippy:
+    name: Cargo clippy
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v2
 
-    - name: fmt
-      run: cargo fmt --all -- --check
+      - name: Install Rust stable toolchain
+        uses: actions-rs/toolchain@v1
+        with:
+            profile: minimal
+            toolchain: stable
+            components: clippy
+            override: true
 
-    - name: build
-      run: cargo build --workspace --verbose
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@v1.3.0
 
-    - name: test
-      run: cargo test --workspace --verbose
+      - name: Run clippy
+        uses: actions-rs/cargo@v1
+        with:
+          command: clippy
+          args: -- -D warnings
```

Comment:
all targets?!

```suggestion
          args: --all-targets -- -D warnings
```

Reply:
Ooh good catch! That picked up some more lint issues to fix too :D :'(

Reply:
maybe postpone it to another PR I suspect that it will plenty of work :)

Reply:
Oh, I've just finished fixing them so it's all gravy :)